crates/ntx-builder/src/actor.rs

```diff
@@ -0,0 +1,248 @@
+use miden_objects::transaction::TransactionId;
+use miden_remote_prover_client::remote_prover::tx_prover::RemoteTransactionProver;
+use tokio::sync::{AcquireError, RwLock, Semaphore, mpsc};
+use tokio_util::sync::CancellationToken;
+use url::Url;
+
+use crate::block_producer::BlockProducerClient;
+use crate::builder::ChainState;
+use crate::state::{State, TransactionCandidate};
+use crate::store::StoreClient;
+use crate::transaction::NtxError;
+
+/// The reason an actor has shut down.
+pub enum ActorShutdownReason {
+    AccountReverted(NetworkAccountPrefix),
+    EventChannelClosed,
+    SemaphoreFailed(AcquireError),
+    Cancelled(NetworkAccountPrefix),
+}
+
+/// Configuration that is required by all account actors.
+#[derive(Debug, Clone)]
+pub struct AccountActorConfig {
+    /// Client for interacting with the store in order to load account state.
+    pub store: StoreClient,
+    /// Address of the block producer gRPC server.
+    pub block_producer_url: Url,
+    /// Address of the remote prover. If `None`, transactions will be proven locally, which is
+    // undesirable due to the performance impact.
+    pub tx_prover_url: Option<Url>,
+    /// The latest chain state that account actors can rely on.
+    pub chain_state: Arc<RwLock<ChainState>>,
+}
+
+/// The origin of the account which the actor will use to initialize the account state.
+#[derive(Debug)]
+pub enum AccountOrigin {
+    /// Accounts that have been created by a transaction.
+    Transaction(Box<Account>),
+    /// Accounts that already exist in the store.
+    Store(NetworkAccountPrefix),
+}
+
+impl AccountOrigin {
+    /// Returns an [`AccountOrigin::Transaction`] if the account is a network account.
+    pub fn transaction(account: &Account) -> Option<Self> {
+        if account.is_network() {
+            Some(AccountOrigin::Transaction(account.clone().into()))
+        } else {
+            None
+        }
+    }
+
+    /// Returns an [`AccountOrigin::Store`].
+    pub fn store(prefix: NetworkAccountPrefix) -> Self {
+        AccountOrigin::Store(prefix)
+    }
+
+    /// Returns the [`NetworkAccountPrefix`] of the account.
+    pub fn prefix(&self) -> NetworkAccountPrefix {
+        match self {
+            AccountOrigin::Transaction(account) => NetworkAccountPrefix::try_from(account.id())
+                .expect("actor accounts are always network accounts"),
+            AccountOrigin::Store(prefix) => *prefix,
+        }
+    }
+}
+
+/// The mode of operation that the account actor is currently performing.
+#[derive(Default, Debug)]
+enum ActorMode {
+    #[default]
+    NoViableNotes,
+    NotesAvailable,
+    TransactionInflight(TransactionId),
+}
+
+/// Account actor that manages state and processes transactions for a single network account.
+pub struct AccountActor {
+    origin: AccountOrigin,
+    store: StoreClient,
+    state: ActorMode,
+    event_rx: mpsc::Receiver<MempoolEvent>,
+    cancel_token: CancellationToken,
```

Comment:
In the doc comments, could we also explain what the `CancellationToken` is used for?